### PR TITLE
Feature/validation on view time page

### PR DESCRIPTION
**Pull request by 7ep**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -226,6 +226,7 @@ task nouitest(type: Test) {
 
 task uitest(type: Test) {
     group = "Verification"
+    maxParallelForks = Runtime.runtime.availableProcessors().intdiv(2) ?: 1
```
Comment (7ep) [inline build.gradle:229]: It's always faster to run the UI tests in parallel.  Not so much the other kinds though.

File: src/main/kotlin/coverosR3z/authentication/persistence/AuthenticationPersistence.kt [hunk 2]
```diff
@@ -9,7 +9,7 @@ import coverosR3z.timerecording.types.EmployeeId
 
 class AuthenticationPersistence(private val pmd : PureMemoryDatabase, private val logger: ILogger) : IAuthPersistence {
 
-    override fun createUser(name: UserName, hash: Hash, salt: Salt, employeeId: EmployeeId?) : User {
+    override fun createUser(name: UserName, hash: Hash, salt: Salt, employeeId: EmployeeId) : User {
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/authentication/persistence/AuthenticationPersistence.kt:12]: all users now have an employee, it's non-nullable (good job @coveros-matt-taylor )

File: src/main/kotlin/coverosR3z/misc/types/Date.kt [hunk 3]
```diff
@@ -9,6 +9,9 @@ import java.text.SimpleDateFormat
 
 const val dateNotNullMsg = "date must not be null"
 const val dateNotBlankMsg = "date must not be blank"
+val earliestAllowableDate = LocalDate.of(1980, 1, 1)
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/misc/types/Date.kt:12]: making these available across the system - so tests can assert we do properly

File: src/main/kotlin/coverosR3z/misc/types/Date.kt [hunk 4]
```diff
@@ -42,17 +45,20 @@ class Date(val epochDay : Int) : Comparable<Date> {
...
 
-    private val sdf: SimpleDateFormat = SimpleDateFormat("MMDDYYYY")
+    /**
+     * See here for the description of this formatting: https://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html
+     */
+    private val sdf: SimpleDateFormat = SimpleDateFormat("MMddYYYY")
 
     /**
-     * Chrome format "MMDDYYYY"
+     * Chrome format "MMddYYYY"
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/misc/types/Date.kt:54]: found a little bug - uppercase D is the number of the day in the year, not the month

File: src/main/kotlin/coverosR3z/server/utility/ServerUtilities.kt [hunk 5]
```diff
@@ -114,6 +115,13 @@ class ServerUtilities() {
                 // without getting anything.  See SocketWrapper and soTimeout
                 logger.logTrace { "read timed out" }
                 shouldKeepAlive = false
+            } catch (ex : SocketException) {
+                if (ex.message?.contains("Connection reset") == true) {
+                    logger.logTrace { "client closed their connection while we were waiting to read from the socket" }
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/server/utility/ServerUtilities.kt:120]: more properly handling the case when a client shuts down their connection to us, we'll pass this up to be handled closer to the top loop of the server

File: src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt [hunk 6]
```diff
@@ -176,10 +178,10 @@ class ViewTimeAPI(private val sd: ServerData) {
                 <input readonly name="${Elements.DATE_INPUT.getElemName()}" type="text" value="${safeAttr(it.date.stringValue)}" />
             </td>
             <td class="time">
-                <input readonly name="${Elements.TIME_INPUT.getElemName()}" type="number" inputmode="decimal" step="0.25"  min="0" max="24" value="${it.time.getHoursAsString()}" />
+                <input readonly name="${Elements.TIME_INPUT.getElemName()}" type="number" value="${it.time.getHoursAsString()}" />
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt:181]: no need for validation on read-only inputs

File: src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt [hunk 7]
```diff
@@ -207,7 +209,7 @@ class ViewTimeAPI(private val sd: ServerData) {
                 </select>
             </td>
             <td class="date">
-                <input name="${Elements.DATE_INPUT.getElemName()}" type="date" value="${safeAttr(it.date.stringValue)}" />
+                <input name="${Elements.DATE_INPUT.getElemName()}" type="date" min="$earliestAllowableDate" max="$latestAllowableDate" value="${safeAttr(it.date.stringValue)}" />
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt:212]: See? We can use constants.  They're nice.

File: src/main/kotlin/coverosR3z/timerecording/types/Employee.kt [hunk 8]
```diff
@@ -45,7 +45,7 @@ data class EmployeeId(val value: Int) {
 
     init {
         require(value < maxEmployeeCount) { maxEmployeeMsg }
-        require(value > 0) { minEmployeeIdMsg }
+        require(value >= 0) { minEmployeeIdMsg }
```
Comment (7ep) [inline src/main/kotlin/coverosR3z/timerecording/types/Employee.kt:48]: now we allow 0 as an employee id - but only for the system user

File: src/test/kotlin/coverosR3z/persistence/PureMemoryDatabaseTests.kt [hunk 9]
```diff
@@ -762,19 +762,6 @@ class PureMemoryDatabaseTests {
         assertEquals(user, deserialized)
     }
 
-    @Test
-    fun testSerialization_UserWithNullEmployee() {
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/persistence/PureMemoryDatabaseTests.kt:766]: obsolete test - no longer possible to have a null employee

File: src/test/kotlin/coverosR3z/server/RegisterAPITests.kt [hunk 10]
```diff
@@ -116,23 +116,6 @@ class RegisterAPITests {
         assertEquals("Must be able to parse $employee as integer", ex.message)
     }
 
-    /**
-     * If the employee id is zero
-     */
-    @Category(APITestCategory::class)
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/server/RegisterAPITests.kt:122]: obsolete

File: src/test/kotlin/coverosR3z/timerecording/EnterTimeAPITests.kt [hunk 11]
```diff
@@ -100,23 +100,6 @@ class EnterTimeAPITests {
         assertEquals("expected keys: [project_entry, time_entry, detail_entry, date_entry]. received keys: [project_entry, time_entry, date_entry]", ex.message)
     }
 
-    /**
-     * If we are missing required data
-     */
-    @Category(APITestCategory::class)
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/timerecording/EnterTimeAPITests.kt:106]: obsolete

File: src/test/kotlin/coverosR3z/timerecording/UITimeEntryTests.kt [hunk 12]
```diff
@@ -57,8 +60,39 @@ class UITimeEntryTests {
      */
     @Category(UITestCategory::class)
     @Test
-    @Ignore("Not started yet")
     fun `all inputs should restrict to valid values` () {
+        val projecta = "projecta"
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/timerecording/UITimeEntryTests.kt:64]: This was a bear!

File: src/test/kotlin/coverosR3z/timerecording/UITimeEntryTests.kt [hunk 13]
```diff
@@ -269,33 +303,16 @@ class UITimeEntryTests {
...
-            DEFAULT_DATE_STRING
-        }
-
-        // Enter time
-        pom.etp.enterTime(project, "1", "", dateString)
+        pom.vtp.enterTime(project, "1", "", pom.calcDateString(DEFAULT_DATE))
     }
 
-
-    private fun calcDateString(date : Date) : String {
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/timerecording/UITimeEntryTests.kt:283]: Moved this to the POM

File: src/test/kotlin/coverosR3z/uitests/UITestHelpers.kt [hunk 14]
```diff
@@ -47,22 +46,6 @@ fun startupTestForUIWithoutServer(domain: String = "", port : Int, driver: () ->
     return PageObjectModel.make(driver(), port, domain)
 }
 
-class EnterTimePage(private val driver: WebDriver, private val domain : String) {
-
-    fun enterTime(project: String, time: String, details: String, date: String) {
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/uitests/UITestHelpers.kt:52]: this is only using view time - not even sure how it landed here.  Moved this to ViewTimePage

File: src/test/kotlin/coverosR3z/uitests/UITestHelpers.kt [hunk 15]
```diff
@@ -201,4 +187,112 @@ class ViewTimePage(private val driver: WebDriver, private val domain: String) {
         val end = driver.findElement(By.name(SubmitTimeAPI.Elements.END_DATE.getElemName())).getAttribute("value")
         return "$start - $end"
     }
+
+    fun enterTime(project: String, time: String, details: String, date: String) {
```
Comment (7ep) [inline src/test/kotlin/coverosR3z/uitests/UITestHelpers.kt:191]: A lot of new capabilities in the ViewTimePage

File: src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt (hunk 7 above)
Comment (coveros-matt-taylor) [reply src/main/kotlin/coverosR3z/timerecording/api/ViewTimeAPI.kt:212]: We could also dynamically change these bounds to only w/in the current period. If we want...

File: src/test/kotlin/coverosR3z/uitests/PageObjectModel.kt [hunk 16]
```diff
@@ -10,14 +12,26 @@ open class PageObjectModel {
...
 
+    /**
+     * Chrome takes date input differently than other browsers.
+     * This helps us avoid a bit of boilerplate so we can use
+     * the proper
+     */
+    fun calcDateString(date : Date) : String {
+        return if (driver is ChromeDriver) {
+            date.chromeStringValue // returns the chrome format
+        } else {
```
Comment (coveros-matt-taylor) [inline src/test/kotlin/coverosR3z/uitests/PageObjectModel.kt:30]: Return _chrome_ format?

**Review by coveros-matt-taylor — CHANGES_REQUESTED** [review]

**Comment by coveros-matt-taylor** [comment]
Looking good!

**Review by coveros-matt-taylor — APPROVED** [review]

**Pull request merged**